### PR TITLE
Add causal support for conv1d

### PR DESCRIPTION
**Pull request by kalyc**
Following implementation similar to https://github.com/dmlc/keras/pull/97/files

@sandeep-krishnamurthy @roywei

### CODE REVIEW

**Comment by sandeep-krishnamurthy** [comment]
@dmadeka

File: keras/backend/mxnet_backend.py [hunk 1]
```diff
@@ -3043,8 +3043,11 @@ def conv1d(x, kernel, strides=1, padding='valid',
 
     # Causal requires temporal padding.
     # MXNet backend does not support temporal padding on 3D tensor.
-    if padding is 'causal':
-        raise ValueError('MXNet Backend: conv1d does not support "causal" padding mode')
+    kernel_shape = kernel.shape
+    if padding == 'causal':
+        pad = dilation_rate * (kernel_shape[0] - 1)
```
Comment (sandeep-krishnamurthy) [inline keras/backend/mxnet_backend.py]: Can you please code comments describing the logic. Will help us later.

File: keras/backend/mxnet_backend.py [hunk 2]
```diff
@@ -3043,8 +3043,11 @@ def conv1d(x, kernel, strides=1, padding='valid',
...
+        x = temporal_padding(x, (pad, 0))
```
Comment (sandeep-krishnamurthy) [inline keras/backend/mxnet_backend.py]: MXNet temporal_padding supports only 3D and 4D tensor. Is it something we need to check here?

Comment (kalyc) [reply keras/backend/mxnet_backend.py]: we check for dimension size inside temporal_padding function - should we throw an error message here?

File: tests/keras/layers/convolutional_test.py [hunk 3]
```diff
@@ -72,8 +72,9 @@ def test_conv_1d():
     input_dim = 2
     kernel_size = 3
     filters = 3
+    paddings = _convolution_paddings + ['causal'] if K.backend() == 'mxnet' else _convolution_paddings
```
Comment (roywei) [inline tests/keras/layers/convolutional_test.py]: shall we change to logic to: add 'causal' to _convolution_paddings if K.backend() != 'theano' ? Now only theano does not support causal padding.

File: keras/backend/mxnet_backend.py (hunk 2 above)
Comment (sandeep-krishnamurthy) [reply keras/backend/mxnet_backend.py]: I would check here can be handled more meaningfully. Because when user adds a Conv1D layer and gets temporal padding error, may not be directly telling what theissue is

File: keras/backend/mxnet_backend.py [hunk 4]
```diff
@@ -3043,8 +3043,11 @@ def conv1d(x, kernel, strides=1, padding='valid',
...
+        padding = 'valid'
 
     if padding not in {'same', 'valid'}:
```
[the remark this answers is not in the document]
Comment (kalyc) [reply keras/backend/mxnet_backend.py]: added

File: keras/backend/mxnet_backend.py (hunk 1 above)
Comment (kalyc) [reply keras/backend/mxnet_backend.py]: done

File: keras/backend/mxnet_backend.py (hunk 2 above)
Comment (kalyc) [reply keras/backend/mxnet_backend.py]: added check for dimensions inside conv1d

File: tests/keras/layers/convolutional_test.py (hunk 3 above)
Comment (kalyc) [reply tests/keras/layers/convolutional_test.py]: Addressed comment
Didn't add this earlier as this was also modifying the way tensorflow unit tests causal conv1d

**Comment by kalyc** [comment]
Tested with `imdb_cnn.py` example - modified padding to be 'causal'
Test results - 
`25000/25000 [==============================] - 468s 19ms/step - loss: 0.4066 - acc: 0.8009 - val_loss: 0.2832 - val_acc: 0.8816`

File: keras/backend/mxnet_backend.py [hunk 5]
```diff
@@ -3042,13 +3042,19 @@ def conv1d(x, kernel, strides=1, padding='valid',
...
     _validate_data_format(data_format)
 
+    if padding not in {'same', 'valid', 'causal'}:
+        raise ValueError('MXNet Backend: `padding` should be either `same`, `valid` or `causal`.')
+
     # Causal requires temporal padding.
     # MXNet backend does not support temporal padding on 3D tensor.
-    if padding is 'causal':
-        raise ValueError('MXNet Backend: conv1d does not support "causal" padding mode')
+    assert ndim(x) == 3
```
Comment (sandeep-krishnamurthy) [inline keras/backend/mxnet_backend.py]: Please add error message about the details in case this assert fails

Comment (kalyc) [reply keras/backend/mxnet_backend.py]: Added
